### PR TITLE
feat(kernel): add SM120 FlashInfer MXFP4 MoE

### PR DESCRIPTION
**Pull request by lucifer1004**
## Summary

- register a FlashInfer CUTLASS MXFP8-activation/MXFP4-weight MoE path for NVIDIA SM120 and SM121
- convert TokenSpeed checkpoint weights, scales, biases, and SwiGLU parameters to the layout expected by FlashInfer
- preserve TokenSpeed precomputed routing and expert/tensor parallel metadata
- cover SM120 selection and checkpoint-layout conversion without changing non-NVIDIA registration

## Motivation

DeepGEMM `mega_moe` serves the existing SM100 path, but it is not available on SM120. FlashInfer 0.6.13, already pinned by TokenSpeed, provides the corresponding CUTLASS MXFP8 x MXFP4 fused MoE kernel on SM120/SM121.

## Impact

`flashinfer_cutlass` can now satisfy MXFP4, SwiGLU, precomputed-top-k MoE plans on SM120/SM121, including expert-parallel plans. Existing architectures and dependency pins are unchanged.

## Validation

- `inferlab run --environment tokenspeed -- pytest tokenspeed/tokenspeed-kernel/test/test_kernel_api_selection.py -q -k mxfp4_cutlass` (`2 passed`)
- `inferlab run --environment tokenspeed -- pytest tokenspeed/tokenspeed-kernel/test/test_runtime_callsite_selection.py -q` (`7 passed, 14 skipped`)
- SM120 kernel probe with BF16 and FP16 inputs completed with finite outputs
- `prek run --all-files`

### CODE REVIEW

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        if intermediate_size % 128 != 0 or hidden_size % 128 != 0:
+            raise ValueError(
+                "FlashInfer SM12x MXFP4 MoE requires hidden and intermediate "
+                f"sizes divisible by 128, got {hidden_size} and {intermediate_size}"
+            )
+
+        # TokenSpeed checkpoints store [w1 (gate) | w3 (up)]. FlashInfer's
+        # CUTLASS SwiGLU path consumes [w3 (up) | w1 (gate)].
+        w13_weight = _swap_halves(w.w13_weight.data, 1)
+        w13_scale = _swap_halves(w.w13_weight_scale.data, 1)
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py]: Fixed in e2b90b1. The preprocessor now branches on w13_input_layout: concatenated checkpoints retain the half swap, while interleaved checkpoints use adjacent pair swaps. The same transformation is applied to weights, scales, and bias. The checkpoint-preservation test now covers both layouts, and the focused MXFP4 plus runtime call-site tests pass.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+            raise ValueError(
+                f"input hidden size {x.shape[-1]} exceeds prepared size {hidden_padded}"
+            )
+        if x.shape[-1] < hidden_padded:
+            x = torch.nn.functional.pad(x, (0, hidden_padded - x.shape[-1]))
+
+        x_quant, x_scale = mxfp8_quantize(
+            x,
+            False,
+            alignment=hidden_padded,
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py:234]: Verified that no code change is needed here. FlashInfer v0.6.14 keeps `SF_VEC_SIZE` fixed at 32; `alignment` only rounds `padded_k`. This path has already padded `x` to `hidden_size_padded` (which is required to be divisible by 128), so using that value leaves both the quantized tensor and scale-vector shape unchanged. An SM120 probe comparing 32 vs. 128 produced identical `q` (`[3, 128]`) and `sf` (`[12]`) tensors. See the [Python implementation](https://github.com/flashinfer-ai/flashinfer/blob/v0.6.14/flashinfer/quantization/fp8_quantization.py) and [C++ implementation](https://github.com/flashinfer-ai/flashinfer/blob/v0.6.14/csrc/nv_internal/tensorrt_llm/thop/fp8Quantize.cpp).

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    size = x.shape[dim]
+    if size % 2 != 0:
+        raise ValueError(f"expected even size in dim {dim}, got {size}")
+    if layout == "concatenated":
+        first, second = x.split(size // 2, dim=dim)
+        return torch.cat((second, first), dim=dim).contiguous()
+    if layout == "interleaved":
+        shape = list(x.shape)
+        paired_shape = shape[:dim] + [size // 2, 2] + shape[dim + 1 :]
+        return x.reshape(paired_shape).flip(dim + 1).reshape(shape).contiguous()
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py]: Fixed in d34fb6e. Interleaved checkpoints are now de-interleaved by gathering odd rows (`w3`) before even rows (`w1`), producing the same `[w3_all | w1_all]` layout as concatenated checkpoints. The normalization covers weights, scales, and bias, and the regression test now asserts the target layout for both input formats.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        )
+        w.w2_weight = torch.nn.Parameter(w2_weight, requires_grad=False)
+        w.w2_weight_scale = torch.nn.Parameter(
+            _swizzle_mxfp4_block_scales(w2_scale),
+            requires_grad=False,
+        )
+
+        if hasattr(w, "w13_weight_bias"):
+            w.w13_weight_bias = torch.nn.Parameter(
+                _reorder_w13(w.w13_weight_bias.data, w13_layout, 1),
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py:141]: Verified that this backend has a different bias contract, so no code change is needed. FlashInfer v0.6.14 explicitly checks both expert-bias tensors against the runner output dtype (`CHECK_INPUT_TYPE(..., mOutputDtype)`). This call sets `output_dtype=torch.bfloat16`, so the existing BF16 biases are the required dtype; casting them to FP32 would instead fail that check. The FP32 conversions in the Triton and TRTLLM preprocessors belong to those kernels, not this CUTLASS API. See the [v0.6.14 binding](https://github.com/flashinfer-ai/flashinfer/blob/v0.6.14/csrc/fused_moe/cutlass_backend/flashinfer_cutlass_fused_moe_binding.cu#L280-L282).

**Comment by weimin023** [comment]
Hi @lucifer1004, do you still work on this?

**Comment by lucifer1004** [comment]
Superseded by #992, which rebases the complete SM120 DeepSeek V4 Flash integration on current main and includes the runtime, attention, indexer, communication, and FlashInfer autotuner API changes in one PR.
